src/ripple/basics/join.h

```diff
@@ -0,0 +1,109 @@
+//------------------------------------------------------------------------------
+/*
+This file is part of rippled: https://github.com/ripple/rippled
+Copyright (c) 2022 Ripple Labs Inc.
+
+Permission to use, copy, modify, and/or distribute this software for any
+purpose  with  or without fee is hereby granted, provided that the above
+copyright notice and this permission notice appear in all copies.
+
+THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+
+#ifndef JOIN_H_INCLUDED
+#define JOIN_H_INCLUDED
+
+// Needed for std::begin and end
+#include <array>
+
+namespace ripple {
+
+template <class Stream, class Iter, class Str>
+Stream&
+join(Stream& s, Iter iter, Iter end, Str const& delimiter)
+{
+    if (iter == end)
+        return s;
+    s << *iter;
+    for (++iter; iter != end; ++iter)
+        s << delimiter << *iter;
+    return s;
+}
+
+template <class Collection, class Str>
+class CollectionAndDelimiter
+{
+public:
+    Collection const& collection;
+    Str const& delimiter;
```

Comment:
The lifetime management of `delimiter` gives me pause. In the common case - passing a statically allocated c string - we're fine. But this is a very general class and I could see `delimiter` pointing to some temporary that went out of scope. Since `delimiters` are going to be small (usually one or two characters) and `std::string` has a small-object optimization (no allocations for small strings), what do you think about using a `std::string` for the delimiter instead?

FWIW, I'm _not_ concerned about the lifetime management of `collection`. The caller should expect this to keep a reference to the object. 

I saw in your tests that you were using integers and uint256 as crazy delimiters. But I think just supporting strings is enough (if you have a crazy delimiter, convert it to a string first). Here's a patch if you want to see what this looks like: https://github.com/seelabs/rippled/commit/2bf76479b11f714086726be7d6c639d3f827cc92